torchinductor/lowering.py

```diff
@@ -2526,6 +2543,52 @@ def copy_(dst, src, non_blocking=False):
     return mutate_to(dst, src)
 
 
+@make_pointwise
+def floordiv(a, b):
+    return ops.floordiv(a, b)
+
+
+@make_pointwise
+def truncdiv(a, b):
+    return ops.truncdiv(a, b)
+
+
+@register_lowering(aten.div.Tensor_mode)
+def div_mode(a, b, rounding_mode=None):
+    both_integer = is_integer_type(a) and is_integer_type(b)
+    both_boolean = is_boolean_type(a) and is_boolean_type(b)
+    on_triton = is_triton(a) or is_triton(b)
+
+    # floordiv and truncdiv need special handling for integer tensors on Triton,
+    # see the discussion at https://github.com/openai/triton/issues/605
+    if rounding_mode == "floor":
+        assert not both_boolean, "floordiv operands can not be boolean at the same time"
+        return floordiv(a, b) if (both_integer and on_triton) else floor(div(a, b))
+    if rounding_mode == "trunc":
+        assert not both_boolean, "truncdiv operands can not be boolean at the same time"
+        return truncdiv(a, b) if (both_integer and on_triton) else trunc(div(a, b))
+    return div(a, b)
+
+
+@register_lowering([aten.div, prims.div])
+def div(a, b):
+    def fn(*args):
+        return ops.div(*args)
+
+    dtype = torch.result_type(
```

Comment:
no, this is still wrong. Are `a` and `b` not necessarily tensors or numbers so you cannot call `torch.result_type` directly? Type promotion rules differ for 0d and 1d or more tensors, and here you are creating a 0d tensor instead of whatever tensor actually was, you need faithful tensor models to send to `result_type`

Reply:
Oh, here `a` and `b` are either scalars or IRs (TensorBox, etc.). They are not really torch.tensor at this point. An alternative solution is to do something similar to https://github.com/pytorch/torchdynamo/blob/cf0d110d58e46f75904c07eea2a2f539490aeed7/torchinductor/lowering.py#L119, what do you think?

Reply:
The existing dtype computation is also wrong, and we really should be fixing it everywhere. You can't just reduce the list of dtypes, because dtypes priority varies on whether the argument is a numer, a 0d tensor or a 1d-or-more tensor. Is it possible to get dimensionality of TensorBox?